File: LeetCode 30 Day Challenge/09-backspaceCompare.js
//O(n) Time O(n) Space
var backspaceCompare = function(S, T) {
    return helper(S) === helper(T)
};

const helper = (str) => {
    let result = []
    for (const current of str) {   
        if (current !== '#') {
            result.push(current)
        } else {
            result.pop()
        }
    }
    return result.join('')

}

//O(n) Time O(1) space

File: LeetCode 30 Day Challenge/09-backspace_compare.py
class Solution(object):
    def backspaceCompare(self, S, T):
        return self.helper(S) == self.helper(T)
    
    def helper(self, str):
        result = []
        for char in str:
            if char != '#':
                result.append(char)
            elif result:
                result.pop()
        return ''.join(result)